contracts/FungibleToken-v2.cdc

```diff
@@ -0,0 +1,197 @@
+
+It also gives users the ability to make custom resources that implement
+these interfaces to do various things with the tokens.
+For example, a faucet can be implemented by conforming
+to the Provider interface.
+
+*/
+
+/// FungibleToken
+///
+pub contract FungibleToken {
+
+    /// Provider
+    ///
+    /// The interface that enforces the requirements for withdrawing
+    /// tokens from the implementing type.
+    ///
+    /// It does not enforce requirements on `balance` here,
+    /// because it leaves open the possibility of creating custom providers
+    /// that do not necessarily need their own balance.
+    ///
+    pub resource interface Provider {
+
+        /// withdraw subtracts tokens from the owner's Vault
+        /// and returns a Vault with the removed tokens.
+        ///
+        /// The function's access level is public, but this is not a problem
+        /// because only the owner storing the resource in their account
+        /// can initially call this function.
+        ///
+        /// The owner may grant other accounts access by creating a private
+        /// capability that allows specific other users to access
+        /// the provider resource through a reference.
+        ///
+        /// The owner may also grant all accounts access by creating a public
+        /// capability that allows all users to access the provider
+        /// resource through a reference.
+        ///
+        pub fun withdraw(amount: UFix64): @AnyResource{Vault} {
+            post {
+                // `result` refers to the return value
+                result.getBalance() == amount:
+                    "Withdrawal amount must be the same as the balance of the withdrawn Vault"
+            }
+        }
+    }
+    
+    pub resource interface Transferable {
+        /// Function for a direct transfer instead of having to do a deposit and withdrawal
+        ///
+        pub fun transfer(amount: UFix64, recipient: Address)
+    }
+
+    /// Receiver
+    ///
+    /// The interface that enforces the requirements for depositing
+    /// tokens into the implementing type.
+    ///
+    /// We do not include a condition that checks the balance because
+    /// we want to give users the ability to make custom receivers that
+    /// can do custom things with the tokens, like split them up and
+    /// send them to different places.
+    ///
+    pub resource interface Receiver {
+
+        /// deposit takes a Vault and deposits it into the implementing resource type
+        ///
+        pub fun deposit(from: @AnyResource{Vault})
+
+        /// getAcceptedTypes optionally returns a list of vault types that this receiver accepts
+        pub fun getAcceptedTypes(): [Type]?
+    }
+
+    /// Balance
+    ///
+    pub resource interface Balance {
+
+        /// Method to get the balance
+        /// The balance could be a derived field,
+        /// so there is no need to require an explicit field
+        pub fun getBalance(): UFix64
```

Comment:
Maybe it is because a derived field will make it easy to add some special mechanisms like rebase

Reply:
> @dete Was the one who made the ask for a derived field. Maybe he can provide some insight
> Maybe it is because a derived field will make it easy to add some special mechanisms like rebase

Highly discouraged to use getBalance() on flow just for an rebase token.
Since the contract codes can be updated on flow, malicious project can modify getBalance() to attack contracts that include this function.
It is risk to import an external function only for reading balance. for example, re-entrancy attack could ocuur

Reply:
To give another side of the coin for balance being a function which I ran into recently while implementing [ScopedFTProviders](https://github.com/green-goo-dao/flow-utils/blob/main/cadence/contracts/ScopedFTProviders.cdc)

Because balance is not a function, I could not actually wrap the FT standard because I need to delegate balance to the underlying capability being stored.